From 05a132c42376097c1b1b57c9e23dde8059aa9bf3 Mon Sep 17 00:00:00 2001
From: Jiri Benc <jbenc@redhat.com>
Date: Fri, 12 Sep 2014 14:29:35 -0400
Subject: [net] Add utility function to copy skb hash

Message-id: <9433b5054584835ac7e90d94c76e9cd3dc8eddac.1410531783.git.jbenc@redhat.com>
Patchwork-id: 93990
O-Subject: [RHEL7.1 net PATCH 02/50] net: Add utility function to copy skb hash
Bugzilla: 1110384
RH-Acked-by: Flavio Leitner <fbl@redhat.com>

Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1110384
Brew: https://brewweb.devel.redhat.com/taskinfo?taskID=7948291

This patch was a part of 4-patch series from which, strangely, only two
patches (7539fadcb814 and 09323cc47931) were backported. Bring in the
remaining two.

commit 3df7a74e797aa2d8be9b7c649cfd56a8517dcf6e
Author: Tom Herbert <therbert@google.com>
Date:   Sun Dec 15 22:16:29 2013 -0800

    net: Add utility function to copy skb hash

    Adds skb_copy_hash to copy rxhash and l4_rxhash from one skb to another.

    Signed-off-by: Tom Herbert <therbert@google.com>
    Signed-off-by: David S. Miller <davem@davemloft.net>

Signed-off-by: Jiri Benc <jbenc@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/include/linux/skbuff.h b/include/linux/skbuff.h
index a034381..eaeb9c5 100644
--- a/include/linux/skbuff.h
+++ b/include/linux/skbuff.h
@@ -849,6 +849,12 @@ static inline void skb_clear_hash_if_not_l4(struct sk_buff *skb)
   skb_clear_hash(skb);
 }
 
+static inline void skb_copy_hash(struct sk_buff *to, const struct sk_buff *from)
+{
+ to->rxhash = from->rxhash;
+ to->l4_rxhash = from->l4_rxhash;
+};
+
 #ifdef NET_SKBUFF_DATA_USES_OFFSET
 static inline unsigned char *skb_end_pointer(const struct sk_buff *skb)
 {
diff --git a/net/core/skbuff.c b/net/core/skbuff.c
index 2e2f598..ae34c8c 100644
--- a/net/core/skbuff.c
+++ b/net/core/skbuff.c
@@ -701,9 +701,8 @@ static void __copy_skb_header(struct sk_buff *new, const struct sk_buff *old)
  new->inner_network_header = old->inner_network_header;
  new->inner_mac_header = old->inner_mac_header;
  skb_dst_copy(new, old);
- new->rxhash  = old->rxhash;
+ skb_copy_hash(new, old);
  new->ooo_okay  = old->ooo_okay;
- new->l4_rxhash  = old->l4_rxhash;
  new->no_fcs  = old->no_fcs;
  new->encapsulation = old->encapsulation;
  new->encap_hdr_csum = old->encap_hdr_csum;
-- 
1.7.1